git_machete/cmd.py

```diff
@@ -1094,20 +1104,83 @@ def merge_base(cli_ctxt: CommandLineContext, sha1: str, sha2: str) -> str:
 
 # Note: the 'git rev-parse --verify' validation is not performed in case for either of earlier/later
 # if the corresponding prefix is empty AND the revision is a 40 hex digit hash.
-def is_ancestor(cli_ctxt: CommandLineContext, earlier_revision: str, later_revision: str, earlier_prefix: str = "refs/heads/", later_prefix: str = "refs/heads/") -> bool:
-    if earlier_prefix == "" and is_full_sha(earlier_revision):
-        earlier_sha: Optional[str] = earlier_revision
-    else:
-        earlier_sha = commit_sha_by_revision(cli_ctxt, earlier_revision, earlier_prefix)
-    if later_prefix == "" and is_full_sha(later_revision):
-        later_sha: Optional[str] = later_revision
-    else:
-        later_sha = commit_sha_by_revision(cli_ctxt, later_revision, later_prefix)
+def is_ancestor(
+    cli_ctxt: CommandLineContext,
+    earlier_revision: str,
+    later_revision: str,
+    earlier_prefix: str = "refs/heads/",
+    later_prefix: str = "refs/heads/",
+) -> bool:
+    earlier_sha = full_sha(cli_ctxt, earlier_revision, earlier_prefix)
+    later_sha = full_sha(cli_ctxt, later_revision, later_prefix)
+
     if earlier_sha == later_sha:
         return True
     return merge_base(cli_ctxt, earlier_sha, later_sha) == earlier_sha
 
 
+# Determine if any revisions in later_revision contain a tree with identical
+# contents to earlier_revision, indicating that later_revision contains a
+# rebase or squash merge of earlier_revision.
+def contains_equivalent_tree(
+    cli_ctxt: CommandLineContext,
+    earlier_revision: str,
+    later_revision: str,
+    earlier_prefix: str = "refs/heads/",
+    later_prefix: str = "refs/heads/",
+) -> bool:
+    earlier_sha = full_sha(cli_ctxt, earlier_revision, earlier_prefix)
+    later_sha = full_sha(cli_ctxt, later_revision, later_prefix)
+
+    if earlier_sha == later_sha:
+        return True
+
+    debug(
+        cli_ctxt,
+        ">>> contains_equivalent_tree",
+        f"earlier_revision={earlier_revision} later_revision={later_revision}",
+    )
+
+    # git rev-list later_sha ^earlier_sha
+    # shows all commits reachable via later_sha but not by earlier_sha
+    #
+    # https://git-scm.com/docs/git-rev-list#_description
+    intermediate_shas = [
+        sha.strip()
+        for sha in popen_git(
+            cli_ctxt,
+            "rev-list",
+            later_sha,
+            "^" + earlier_sha,
+        ).splitlines()
+        if sha.strip()
+    ]
+
+    for intermediate_sha in intermediate_shas:
+        # git diff-tree --quiet
+        # Exits with 1 if there were differences and 0 means no differences.
+        #
+        # https://git-scm.com/docs/git-diff-tree#Documentation/git-diff-tree.txt---quiet
+        diff_result = run_cmd(
+            cli_ctxt,
+            "git",
+            "diff-tree",
```

Comment:
Hmm actually... would it make sense to just compare the hashes of trees contained within `earlier_sha` and `intermediate_sha`? 🤔 although I suppose that's exactly what `git diff-tree` does anyway (as in, I don't think it doesn't actually go one-by-one by the files contained within either commit since it doesn't make sense in a content-addressable storage like git has ;) )